Async version of mcp23017's switch and binary_sensor entities

jpcornil-git opened the pull request:
- **Async code** is implemented in entities while sync code ('mcp23017 device driver') is implemented in ```__init__.py``` .
- **Code is threadsafe** and allows different entities to use the same component (fix issue #31867).
- **Config Flow** is now supported in addition to legacy configuration.yaml.
- **Push iso pull model** is used for Async entities, active polling implemented in the sync code.
  - offer much higher reactivity, e.g. 100ms polling for 'zero-delay' push button, without loading HA.
- Dependencies towards all AdaFruit libraries have been removed.
  - mcp23017 integration can now run on a std linux machine (and be tested using i2c_stub module).
- Optimized i2c bus bandwidth by reducing number of transactions (polling per device instead of per entity/8x gain, register cache to avoid read-modify-write/3xgain or rewriting the same register value).
- Synchronization with the device state at startup, e.g. avoid output glitches when HA restart.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
None

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue): fix issue #31867
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

binary_sensor:
  - platform: mcp23017
    i2c_address: 0x26
    pins:
      8 : Button_0
      9 : Button_1
      10: Button_2
      11: Button_3
  - platform: mcp23017
    i2c_address: 0x27
    invert_logic: true
    pins:
      8: Button_4
      9: Button_5
      10: Button_6
      11: Button_7

switch:
  - platform: mcp23017
    i2c_address: 0x26
    pins:
      0 : Output_0
      1 : Output_1
      2 : Output_2
      3 : Output_3
  - platform: mcp23017
    i2c_address: 0x27
    pins:
      0 : Output_4
      1 : Output_5
      2 : Output_6
      3 : Output_7
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #31867
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jpcornil-git committed:
Async version of mcp23017's switch and binary_sensor entities

- Async code implemented in entities, sync code implemented in __init__.py (mcp23017 device driver)
- Code is now threadsafe and authorize different entities to use the same component
- Push model is now used for Async entities, active polling implemented in the sync code (much higher reactivity)
- Synchronization with the device at startup (avoid output glitches when HA restart)
- New i2c component created to enable registration of other integrations in a common framework and remove dependencies towards all AdaFruit libraries (only std smbus2, i.e. can also run on a std linux machine and be tested i2c_stub module)
- Optimized i2c bus bandwidth by reducing number of transactions (polling per device instead of per entity/8x gain, register cache to avoid read-modify-write/3xgain)

jpcornil-git committed:
Fixing pylint error

jpcornil-git committed:
Add missing i2c dependency

suba83 commented:
Hellow, i have problem with integration "mcp23017" to Home Assistant (117.6). When i added to configuration file 
switch: 
- platform: mcp23017 ...
i have error:

Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 418, in start
    resp = await task
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_app.py", line 458, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_middlewares.py", line 119, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 18, in request_context_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 72, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 127, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 129, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/core.py", line 29, in post
    errors = await async_check_ha_config_file(request.app["hass"])
  File "/usr/src/homeassistant/homeassistant/config.py", line 874, in async_check_ha_config_file
    res = await check_config.async_check_ha_config_file(hass)
  File "/usr/src/homeassistant/homeassistant/helpers/check_config.py", line 164, in async_check_ha_config_file
    platform = p_integration.get_platform(domain)
  File "/usr/src/homeassistant/homeassistant/loader.py", line 422, in get_platform
    cache[full_name] = self._import_platform(platform_name)
  File "/usr/src/homeassistant/homeassistant/loader.py", line 427, in _import_platform
    return importlib.import_module(f"{self.pkg_path}.{platform_name}")
  File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/src/homeassistant/homeassistant/components/mcp23017/binary_sensor.py", line 3, in <module>
    import board  # pylint: disable=import-error
  File "/usr/local/lib/python3.8/site-packages/board.py", line 181, in <module>
    raise NotImplementedError("Board not supported {}".format(board_id))
NotImplementedError: Board not supported None

Can you help me?

jpcornil-git commented:
The current mcp23017 implementation depends upon and external `adafruit-circuitpython-mcp230xx` package that depends itself upon `Adafruit_Blinka` package that includes the `board `module.

Reported error (cfr. L181 of https://github.com/adafruit/Adafruit_Blinka/blob/master/src/board.py) seems to indicate that either your platform (RPi4 ?) is not supported or something has been changed in one of these packages and broke dependencies again  (recurring issue/why this PR also remove all these dependencies).

You should probably open/report an issue rather than use this PR (and I can eventually help you there)

Cheers,
jpc

suba83 commented:
> The current mcp23017 implementation depends upon and external `adafruit-circuitpython-mcp230xx` package that depends itself upon `Adafruit_Blinka` package that includes the `board `module.
> 
> Reported error (cfr. L181 of https://github.com/adafruit/Adafruit_Blinka/blob/master/src/board.py) seems to indicate that either your platform (RPi4 ?) is not supported or something has been changed in one of these packages and broke dependencies again (recurring issue/why this PR also remove all these dependencies).
> 
> You should probably open/report an issue rather than use this PR (and I can eventually help you there)
> 
> Cheers,
> jpc

Thank you!

jpcornil-git committed:
Add support for config flow

jpcornil-git committed:
Fix backward compatibility issue

jpcornil-git committed:
- Fix a few pylint warnings
- Remove a few unused function parameters
- Update MCP23017 output when invert_logic configuration is changed
- Deferred object initialization (blocking calls) to the thread pool as well

jpcornil-git committed:
Add pytest support for config flow

jpcornil-git committed:
Mock i2c setup

jpcornil-git commented:
I'm done with this one for the time being/waiting for a review now.
Cheers
jpc

jpcornil-git commented:
@jardiamj, any feedback on this ?

agners commented:
@jpcornil-git don't think that adding a I2C library as a integration is a good idea. Integrations are meant to integrate devices with Home Assistant Core. You are creating an integration to use it as a base library... Is there any reason this is part of HA at all? This should be solved in a library outside of HA.

jpcornil-git commented:
@agners proposed I2C implementation is very HA-specific and can be seen as a 'DataUpdateCoordinator' for I2C device.

I can move all i2c code within existing mcp23017 component but decided to isolate it as I developed other custom components, sharing the same i2c bus and wanted to avoid code replication, multiple threads running the same code and share a common/generic bus locking mechanism for all devices, e.g. e.g. hts221/lps22hb/veml6075 humidity/pressure/uv sensors in https://github.com/jpcornil-git/core/tree/Arduino_MKR_ENV/homeassistant/components.

This architecture also allows multiple mcp23017 to be instantiated on the same bus without threading issues.

agners commented:
> @agners proposed I2C implementation is very HA-specific and can be seen as a 'DataUpdateCoordinator' for I2C device.

Ok, I see your implementation ended up being HA-specifc. But the problem seems rather a generic problem: You need to be able to synchronize an operation on a I2C device between two threads. It should be possible to implement that in a generic manner, e.g. by having some locking mechanism on the I2C device class.

> 
> I can move all i2c code within existing mcp23017 component but decided to isolate it as I developed other custom components, sharing the same i2c bus and wanted to avoid code replication, multiple threads running the same code and share a common/generic bus locking mechanism for all devices, e.g. e.g. hts221/lps22hb/veml6075 humidity/pressure/uv sensors in https://github.com/jpcornil-git/core/tree/Arduino_MKR_ENV/homeassistant/components.

No, moving i2c to the existing mcp23017 seems the wrong way to go to me as well.

> 
> This architecture also allows multiple mcp23017 to be instantiated on the same bus without threading issues.

Yes, that is nice, and I think you are on the right track to synchronize that stuff properly. It just should be solved somewhere else. In this case, I think we need to fix it in the Adafruit libraries. 

From what I can see, the `Adafruit_CircuitPython_MCP230xx` library is based on `Adafruit_Blinka` (and that is based on `Adafruit_PureIO`). I think we should extend the `I2C` class in the Blinka library. I suggest to open a issue on their end referencing our issue, and pointing out the problem (synchronization across multiple threads). Then we can come up with a proper solution with them.

jpcornil-git commented:
I'll review the Adafruit code again to see if synchronization can be fixed there without too much refactoring but even then, to convert this component to async and have an efficient/generic implementation, I need to have a thread/coordinator running somewhere polling i2c devices and pushing updates towards related entities. Polling from async is of course possible but can't be done at high rate (critical when mcp23017 is used as push-button for e.g. lights, you need sub-second polling to have a good user experience)  + ineffective without a coordinator (without a coordinator, you need to poll 16 times iso 1 for a device like mcp23017).

-> Let me come back/refresh my mind on the Adafruit code :-)

jpcornil-git commented:
@agners I went to the Adafruit libraries over the WE and they only offer an alternative way to wrap i2c accesses like the standard python smbus2 library for i2c (but with more dependencies) and not helping with this specific problem.

The main issue to solve here is to allow thread-safe accesses from different entities (sensor, switch, binary_sensor, ...) to common registers within a single device. There should therefore be one object per device to serialize (and optimize) accesses from HA thread pool/executor jobs to the device and post consistent state changes.

What I can do to avoid another associated integration (i2c here) is to limit changes to this component only (PR #44238 was another one but is now merged into dev) by moving the registration/hardware scanning code to the mpc23017 component itself; only drawback is that other i2c components may have to replicate this moving forward but there won't be threading issues between them (linux driver serializing accesses to the bus).

I should be able to do that pretty quickly.

jpcornil-git committed:
Match original commit in upstream/dev branch

jpcornil-git committed:
Remove i2c integration

jpcornil-git committed:
Fix config flow test (remove i2c dependencies)

jpcornil-git commented:
@agners Just checked in a version that removes the i2c integration (label can therefore be removed) and moves the polling code thread within MCP23017 object

jpcornil-git committed:
Fix error handling when an invalid/unavailable i2c is configured
Fix reload of config entry options (was always using default)

jpcornil-git committed:
Fix two pylint warnings

jpcornil-git committed:
Merge branch 'dev' into mcp23017_async

jpcornil-git committed:
Fix new R1729 pylint warnings

jpcornil-git commented:
@jardiamj I'm done with this code refactoring (add async/config flow @MartinHjelmare FI) and bug fix (fix threading issue #31867)

Can someone also remove the 'integration:i2c' and 'new integration' tags (not applicable anymore since I removed that integration to simplify this PR after a discussion with @agners here above)

MartinHjelmare commented on homeassistant/components/mcp23017/__init__.py:
```diff
@@ -1,3 +1,372 @@
+    return component
+
+
+class MCP23017(threading.Thread):
```
This class looks too low level for inclusion in Home Assistant directly. We want to use a library for device integration code.

jpcornil-git replied:
@MartinHjelmare I do agree we you in general but in this case, it would basically mean to remove 7 lines (dict setter/getter L220-L227) and replace statements like:

`self[GPIOA] = value`

by:

`self._device.gpioa(value)`

Everything else would stay the same, i.e.  comparable readability but no dependencies to foreign libraires (adafruit-circuitpython-mcp230xx, Adafruit-Blinka, adafruit-circuitpython-busdevice and assosciated maintenance issues like #31719, #40192) that are not abstracting anything or adding value in this context.

jpcornil-git commented:
@jardiamj are you still maintaining this component ?
No reactions on this PR nor recent #48106 commit made by @frenck ?

coryshack commented:
I have been going over it yes. Learning and I know the direction it needs
to go in and how to get it there but I don’t know the codes specifically.

On Mon, Mar 29, 2021 at 10:48 AM jpcornil-git ***@***.***>
wrote:

> @jardiamj <https://github.com/jardiamj> are you still maintaining this
> component ?
> No reactions on this PR nor recent #48106
> <https://github.com/home-assistant/core/pull/48106> commit made by @frenck
> <https://github.com/frenck> ?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/42928#issuecomment-809440839>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASR3YZRHVKAAATKV5GYYZALTGCOMVANCNFSM4TNQQ2TQ>
> .
>

jardiamj commented:
> @jardiamj are you still maintaining this component ?
> No reactions on this PR nor recent #48106 commit made by @frenck ?

I have been extremely busy lately. I decided to pursue a career in Software Development and got accepted in a program at UW, so I am now working full time and studying Software Engineering.
I am still keeping an eye on this but I haven't had the time to take a careful look. I am one quarter away from finishing the program so I just need to survive 3 more months, but until then I wont have much free time.

coryshack commented:
Sounds exactly like the boat I’m in.

On Mon, Mar 29, 2021 at 11:28 AM Jardi Martinez ***@***.***>
wrote:

> @jardiamj <https://github.com/jardiamj> are you still maintaining this
> component ?
> No reactions on this PR nor recent #48106
> <https://github.com/home-assistant/core/pull/48106> commit made by @frenck
> <https://github.com/frenck> ?
>
> I have been extremely busy lately. I decided to pursue a career in
> Software Development and got accepted in a program at UW, so I am now
> working full time and studying Software Engineering.
> I am still keeping an eye on this but I haven't had the time to take a
> careful look. I am one quarter away from finishing the program so I just
> need to survive 3 more months, but until then I wont have much free time.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/42928#issuecomment-809474144>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASR3YZTLZSZZJUIQHCPGETDTGCMADANCNFSM4TNQQ2TQ>
> .
>

jpcornil-git commented:
OK, parking this for 3 months then

MartinHjelmare commented:
I'll close here then. You're welcome to open again later with a library approach.

jpcornil-git commented:
OK/will leave it there then as I don't believe it is the right approach here (will keep my fork alive a few more months in case someone want to reuse async/config flow implementations or just make use of it to fix threading or latency issues)  
Cheers
jpc

gralin commented:
@jpcornil-git would you mind enabling Discussion or Issues in your repository so that I could ask you questions when I do my tests? My setup with MCP23017 is a bit non-standard and I needed to use a custom component which is a modified version of the existing one to handle it. So in case I also need to modify your version, maybe would be good to discuss it and find a way it could be supported out of the box. This way, if your changes end up in HA one day, I will be happy to remove my custom component (which doesn't have nice async operations like you have implemented).

jpcornil-git commented:
@gralin just done; moving forward, I'll probably move mine to a custom or hacs component to ease maintenance of people already using it today (current one way too slow for e.g. light switches on top of #31867)
Cheers
jpc

jpcornil-git commented:
@gralin, I just created https://github.com/jpcornil-git/HA-mcp23017 to offer it as a "custom_components" (same component name -> it will take precedence over the legacy implementation once installed and you can easily revert back by just removing it from your custom_components directory). Your mcp23017 configuration.yaml should be imported automatically and be available  as if it was created using config flow.

Please use that repository for discussions as well (I'll remove the existing fork over the WE probably)
